File: apps/vis_app/views/api/scitex_editor.py
"""
SciTeX Visual Editor API - Real-time figure editing endpoints.

Integrates scitex.vis functionality into the Django /vis/ page:
- Load figure from JSON/CSV files
- Render preview with overrides
- Save manual edits to .manual.json
"""

import json
from pathlib import Path

from django.http import HttpResponse, JsonResponse
from django.views.decorators.csrf import csrf_exempt
from django.views.decorators.http import require_http_methods

from .editor_render import render_export_figure, render_figure_preview
from .editor_styles import (
    compute_file_hash,
    extract_defaults_from_metadata,
    get_scitex_defaults,
)


@require_http_methods(["POST"])
@csrf_exempt
def load_figure_json(request):
    """Load a figure from JSON file and return metadata + defaults."""
    try:
        data = json.loads(request.body)
        json_path = Path(data.get("json_path", ""))

        if not json_path.exists():
            return JsonResponse(
                {"error": f"JSON file not found: {json_path}"}, status=404
            )

        with open(json_path, "r") as f:
            metadata = json.load(f)

        csv_path, csv_data, csv_columns = _load_csv_data(
            json_path, data.get("csv_path")
        )

        overrides = get_scitex_defaults()
        overrides.update(extract_defaults_from_metadata(metadata))

        manual_path = json_path.with_suffix(".manual.json")
        if manual_path.exists():
            with open(manual_path, "r") as f:
                manual_overrides = json.load(f).get("overrides", {})
            overrides.update(manual_overrides)

        preview = render_figure_preview(metadata, csv_data, overrides)

        return JsonResponse(
            {
                "success": True,
                "metadata": metadata,
                "overrides": overrides,
                "csv_columns": csv_columns,
                "csv_path": str(csv_path) if csv_path else None,
                "json_path": str(json_path),
                "preview": preview,
            }
        )
    except json.JSONDecodeError:
        return JsonResponse({"error": "Invalid JSON data"}, status=400)
    except Exception as e:
        return JsonResponse({"error": f"Failed to load figure: {str(e)}"}, status=500)


@require_http_methods(["POST"])
@csrf_exempt
def update_preview(request):
    """Update figure preview with new overrides."""
    try:
        data = json.loads(request.body)
        json_path_str = data.get("json_path", "")
        json_path = Path(json_path_str) if json_path_str else None
        overrides = data.get("overrides", {})
        sample_data_str = data.get("sample_data")

        metadata = {}
        if json_path and json_path.exists():
            with open(json_path, "r") as f:
                metadata = json.load(f)

        csv_data = None
        if sample_data_str:
            from io import StringIO

            import pandas as pd

            csv_data = pd.read_csv(StringIO(sample_data_str))
        elif data.get("csv_path"):
            import pandas as pd

            csv_path = Path(data.get("csv_path"))
            if csv_path.exists():
                csv_data = pd.read_csv(csv_path)

        full_overrides = get_scitex_defaults()
        full_overrides.update(extract_defaults_from_metadata(metadata))
        full_overrides.update(overrides)

        preview = render_figure_preview(metadata, csv_data, full_overrides)
        return JsonResponse({"preview": preview, "status": "updated"})
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@require_http_methods(["POST"])
@csrf_exempt
def save_manual_overrides(request):
    """Save manual overrides to .manual.json file."""
    try:
        data = json.loads(request.body)
        json_path = Path(data.get("json_path", ""))
        overrides = data.get("overrides", {})

        if not json_path.exists():
            return JsonResponse(
                {"error": f"JSON file not found: {json_path}"}, status=404
            )

        manual_data = {
            "base_file": json_path.name,
            "base_hash": compute_file_hash(json_path),
            "overrides": overrides,
        }

        manual_path = json_path.with_suffix(".manual.json")
        with open(manual_path, "w") as f:
            json.dump(manual_data, f, indent=2)

        return JsonResponse({"status": "saved", "path": str(manual_path)})
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@require_http_methods(["POST"])
@csrf_exempt
def export_figure(request):
    """Export figure in specified format."""
    try:
        data = json.loads(request.body)
        json_path = Path(data.get("json_path", ""))
        overrides = data.get("overrides", {})
        export_format = data.get("format", "png").lower()
        export_dpi = data.get("dpi", 300)

        valid_formats = ["png", "pdf", "svg", "tiff"]
        if export_format not in valid_formats:
            return JsonResponse(
                {"error": f"Invalid format. Valid: {', '.join(valid_formats)}"},
                status=400,
            )

        metadata = {}
        if json_path.exists():
            with open(json_path, "r") as f:
                metadata = json.load(f)

        csv_data = None
        if data.get("csv_path"):
            import pandas as pd

            csv_path = Path(data.get("csv_path"))
            if csv_path.exists():
                csv_data = pd.read_csv(csv_path)

        full_overrides = get_scitex_defaults()
        full_overrides.update(extract_defaults_from_metadata(metadata))
        full_overrides.update(overrides)
        full_overrides["dpi"] = export_dpi

        export_data = render_export_figure(
            csv_data, full_overrides, export_format, export_dpi
        )

        content_types = {
            "png": "image/png",
            "pdf": "application/pdf",
            "svg": "image/svg+xml",
            "tiff": "image/tiff",
        }
        response = HttpResponse(export_data, content_type=content_types[export_format])
        response["Content-Disposition"] = (
            f'attachment; filename="{json_path.stem}.{export_format}"'
        )
        return response
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


@require_http_methods(["GET"])
def get_scitex_style(request):
    """Get SCITEX_STYLE configuration for the frontend."""
    try:
        defaults = get_scitex_defaults()
        return JsonResponse({"style": defaults, "defaults": defaults})
    except Exception as e:
        return JsonResponse({"error": str(e)}, status=500)


def _load_csv_data(json_path, csv_path_str):
    """Load CSV data from path or auto-detect."""
    import pandas as pd

    csv_path = None
    csv_data = None
    csv_columns = []

    if csv_path_str:
        csv_path = Path(csv_path_str)
    else:
        csv_sibling = json_path.with_suffix(".csv")
        if csv_sibling.exists():
            csv_path = csv_sibling
        elif json_path.parent.name == "json":
            csv_organized = json_path.parent.parent / "csv" / f"{json_path.stem}.csv"
            if csv_organized.exists():
                csv_path = csv_organized

    if csv_path and csv_path.exists():
        csv_data = pd.read_csv(csv_path)
        csv_columns = csv_data.columns.tolist()

    return csv_path, csv_data, csv_columns

File: apps/vis_app/static/vis_app/ts/vis/SciTeXEditor.ts
/**
 * SciTeXEditor - Visual figure editor with real-time preview
 *
 * Integrates scitex.vis functionality for:
 * - Loading figures from JSON/CSV files
 * - Real-time preview updates
 * - Property editing (labels, traces, legend, ticks, style, dimensions)
 * - Non-destructive edits saved to .manual.json
 *
 * Refactored: PropertyPanelRenderer handles panel HTML and events.
 */

import type {
    SciTeXFigureMetadata,
    SciTeXFigureOverrides,
    SciTeXTraceConfig,
    SciTeXAnnotationConfig
} from './types';
import { PropertyPanelRenderer } from './PropertyPanelRenderer';

// Re-export types for backwards compatibility
export type FigureMetadata = SciTeXFigureMetadata;
export type FigureOverrides = SciTeXFigureOverrides;
export type TraceConfig = SciTeXTraceConfig;
export type AnnotationConfig = SciTeXAnnotationConfig;

export class SciTeXEditor {
    private containerEl: HTMLElement | null = null;
    private previewEl: HTMLImageElement | null = null;
    private propertiesEl: HTMLElement | null = null;
    private panelRenderer: PropertyPanelRenderer | null = null;

    // Current state
    private jsonPath: string | null = null;
    private csvPath: string | null = null;
    private metadata: FigureMetadata = {};
    private overrides: FigureOverrides = {};
    private isLoading: boolean = false;

    // Callbacks
    private onUpdateCallback?: (overrides: FigureOverrides) => void;

    constructor(options: {
        containerId?: string;
        previewId?: string;
        propertiesId?: string;
        onUpdate?: (overrides: FigureOverrides) => void;
    } = {}) {
        this.containerEl = document.getElementById(options.containerId || 'canvas-container');
        this.previewEl = document.getElementById(options.previewId || 'figure-preview') as HTMLImageElement;
        this.propertiesEl = document.getElementById(options.propertiesId || 'dynamic-properties');
        this.onUpdateCallback = options.onUpdate;

        this.initializeUI();
    }

    /**
     * Initialize the editor UI
     */
    private initializeUI(): void {
        // Create preview image if not exists
        if (this.containerEl && !this.previewEl) {
            this.previewEl = document.createElement('img');
            this.previewEl.id = 'figure-preview';
            this.previewEl.className = 'scitex-figure-preview';
            this.previewEl.alt = 'Figure Preview';
            this.previewEl.style.cssText = `
                max-width: 100%;
                max-height: 100%;
                display: block;
                margin: auto;
            `;
            this.containerEl.appendChild(this.previewEl);
        }

        console.log('[SciTeXEditor] UI initialized');
    }

    /**
     * Load a figure from JSON file path
     */
    public async loadFigure(jsonPath: string, csvPath?: string): Promise<void> {
        if (this.isLoading) {
            console.warn('[SciTeXEditor] Already loading a figure');
            return;
        }

        this.isLoading = true;
        this.setStatus('Loading figure...');

        try {
            const response = await fetch('/vis/api/editor/load/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    json_path: jsonPath,
                    csv_path: csvPath || null,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.error || 'Failed to load figure');
            }

            const data = await response.json();

            this.jsonPath = data.json_path;
            this.csvPath = data.csv_path;
            this.metadata = data.metadata;
            this.overrides = data.overrides;

            // Update preview
            if (data.preview && this.previewEl) {
                this.previewEl.src = `data:image/png;base64,${data.preview}`;
            }

            // Render properties panel
            this.renderPropertiesPanel();

            this.setStatus(`Loaded: ${this.metadata.id || 'figure'}`);
            console.log('[SciTeXEditor] Figure loaded:', this.metadata.id);
        } catch (error) {
            console.error('[SciTeXEditor] Load error:', error);
            this.setStatus(`Error: ${error instanceof Error ? error.message : 'Unknown error'}`, true);
        } finally {
            this.isLoading = false;
        }
    }

    /**
     * Update preview with current overrides
     */
    public async updatePreview(): Promise<void> {
        if (!this.jsonPath) {
            console.warn('[SciTeXEditor] No figure loaded');
            return;
        }

        this.setStatus('Updating...');

        try {
            const response = await fetch('/vis/api/editor/preview/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    json_path: this.jsonPath,
                    csv_path: this.csvPath,
                    overrides: this.overrides,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.error || 'Failed to update preview');
            }

            const data = await response.json();

            if (data.preview && this.previewEl) {
                this.previewEl.src = `data:image/png;base64,${data.preview}`;
            }

            this.setStatus('Preview updated');
            this.onUpdateCallback?.(this.overrides);
        } catch (error) {
            console.error('[SciTeXEditor] Update error:', error);
            this.setStatus(`Error: ${error instanceof Error ? error.message : 'Unknown error'}`, true);
        }
    }

    /**
     * Save manual overrides to .manual.json
     */
    public async saveManualOverrides(): Promise<void> {
        if (!this.jsonPath) {
            console.warn('[SciTeXEditor] No figure loaded');
            return;
        }

        this.setStatus('Saving...');

        try {
            const response = await fetch('/vis/api/editor/save/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    json_path: this.jsonPath,
                    overrides: this.overrides,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.error || 'Failed to save');
            }

            const data = await response.json();
            this.setStatus(`Saved: ${data.path.split('/').pop()}`);
            console.log('[SciTeXEditor] Saved to:', data.path);
        } catch (error) {
            console.error('[SciTeXEditor] Save error:', error);
            this.setStatus(`Error: ${error instanceof Error ? error.message : 'Unknown error'}`, true);
        }
    }

    /**
     * Export figure in specified format
     */
    public async exportFigure(format: 'png' | 'pdf' | 'svg' | 'tiff' = 'png', dpi: number = 300): Promise<void> {
        if (!this.jsonPath) {
            console.warn('[SciTeXEditor] No figure loaded');
            return;
        }

        this.setStatus(`Exporting ${format.toUpperCase()}...`);

        try {
            const response = await fetch('/vis/api/editor/export/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    json_path: this.jsonPath,
                    csv_path: this.csvPath,
                    overrides: this.overrides,
                    format: format,
                    dpi: dpi,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.error || 'Failed to export');
            }

            // Download file
            const blob = await response.blob();
            const url = window.URL.createObjectURL(blob);
            const a = document.createElement('a');
            a.href = url;
            a.download = `${this.metadata.id || 'figure'}.${format}`;
            document.body.appendChild(a);
            a.click();
            document.body.removeChild(a);
            window.URL.revokeObjectURL(url);

            this.setStatus(`Exported: ${a.download}`);
        } catch (error) {
            console.error('[SciTeXEditor] Export error:', error);
            this.setStatus(`Error: ${error instanceof Error ? error.message : 'Unknown error'}`, true);
        }
    }

    /**
     * Render the properties panel with SciTeX editor controls
     */
    private renderPropertiesPanel(): void {
        if (!this.propertiesEl) return;

        // Create or update panel renderer
        if (!this.panelRenderer) {
            this.panelRenderer = new PropertyPanelRenderer(
                this.propertiesEl,
                this.overrides,
                {
                    onUpdatePreview: () => this.updatePreview(),
                    onSave: () => this.saveManualOverrides(),
                    onReset: () => {
                        if (this.jsonPath && confirm('Reset all changes to original values?')) {
                            this.loadFigure(this.jsonPath, this.csvPath || undefined);
                        }
                    }
                }
            );
        } else {
            this.panelRenderer.setOverrides(this.overrides);
        }
        this.panelRenderer.render();
    }

    /**
     * Collect current override values from panel
     */
    private collectOverrides(): void {
        if (this.panelRenderer) {
            this.overrides = this.panelRenderer.collectOverrides();
        }
    }

    /**
     * Set status message
     */
    private setStatus(message: string, isError: boolean = false): void {
        // Find or create status element
        let statusEl = document.getElementById('scitex-status');
        if (!statusEl && this.propertiesEl) {
            statusEl = document.createElement('div');
            statusEl.id = 'scitex-status';
            statusEl.className = 'scitex-status';
            this.propertiesEl.appendChild(statusEl);
        }

        if (statusEl) {
            statusEl.textContent = message;
            statusEl.classList.toggle('error', isError);
        }

        console.log(`[SciTeXEditor] ${isError ? 'ERROR: ' : ''}${message}`);
    }

    /**
     * Get current overrides
     */
    public getOverrides(): FigureOverrides {
        return { ...this.overrides };
    }

    /**
     * Get current metadata
     */
    public getMetadata(): FigureMetadata {
        return { ...this.metadata };
    }

    /**
     * Check if a figure is loaded
     */
    public isLoaded(): boolean {
        return this.jsonPath !== null;
    }
}
